adoptingerlang.org

```diff
@@ -1334,6 +1334,370 @@ Repeat this exercise from time to time, and eventually do some [[https://princip
+- =rebar3 tree= will print out the dependency tree that was built and is represented by the lock file
+- =rebar3 deps= will list out dependencies and annotate those that could be updated. Do note that there are strong limitations in what it considers worth updating: a branch in git, a tag whose reference shifted, or hex versions that weren't specified. But if you specified a package to have version ="1.2.3"= and =1.2.4= is available, it won't tell you anything
+
+There are other commands that are a bit more drastic, namely =rebar3 unlock= and =rebar3 upgrade= (without any arguments). These will just get rid of the lock file for the next build of the project. But overall, all these commands will do everything you need to manage most dependencies.
+
+In general your workflow might look like this:
+
+1. set up the initial project, compile once, and track the lock file in source control
+2. find out you have a top-level dependency you want to change
+3. change the dependency definition in the rebar.config file (or optionally if using a non-versioned hex package, call =rebar3 update=)
+4. call =rebar3 upgrade <app>= to update the application and its transitive dependencies
+
+And that's it, you're done.
+
+*** Checkout Dependencies
+
+If you're just adopting Erlang, whether for fun or at work, chances are you'll have one project to maintain, maybe a few more than that, and small ones at that. Working with a few repositories won't be too hard. But sooner or later, if you have to start patching dependencies or if you're working in a corporate environment with dozens and dozens of repositories, working with rebar3 might become frustrating.
+
+The problem would be that every time you want to try and modify a dependency to do something that its parent application needs, you'll have to commit and publish the change somewhere rebar3 can fetch it, because it will otherwise keep trying to make a repeatable build by fetching remote resources without your fix, or won't recompile things it believes are up to date. This can become annoying really fast when you have to work across repository boundaries and just want to test a change.
+
+So there's a little trickshot of a feature called _checkout dependencies_. Checkout dependencies work as follows:
+
+- you have your main project somewhere on your file system
+- the dependency is declared in the rebar.config file of the main project
+- you also have that dependency somewhere on your file system, as a standalone project
+- you add a =_checkouts/= directory to the main project
+- you either copy or symlink the dependency's directory in the =_checkouts/= directory
+
+From that point on, every time the main app is built, it will add an =ebin/= directory to the dependency's directory in checkouts, and re-compile it as if it were a top-level application within the main app.
```

Comment:
```suggestion
From that point on, every time the main app is built, it will add an =ebin/= directory to the dependency's directory in checkouts, and re-compile it as if it were a top-level application within the main project.
```